### PR TITLE
OC-1013: Linked drafts error handling

### PR DESCRIPTION
**Pull request by finlay-jisc**
The purpose of this PR was to allow publications to be linked to draft publications if the author is author or co-author on the linked publication. Additional behaviour has been added to ensure that publications can't exist with invalid links, or be published with a link to a draft.

---

### Acceptance Criteria:

- Whilst a publication is in draft status:
    - If one of the following occurs:
        - A linked draft is deleted
        - The corresponding author is removed as an author on the linked draft
        - Ownership of the draft is changed and the new corresponding author is not a co-author on the linked draft
    - The link is removed, and the linked section shows as incomplete if there are no other links still present.
- Whilst a publication is in locked status::
    - If one of the following occurs:
        - A linked draft is deleted
        - The corresponding author is removed as an author on the linked draft
        - Ownership of the draft is changed and the new corresponding author is not a co-author on the linked draft
    - The link is removed, and the publication is returned to draft status with the linked publications section incomplete.
- During publication, a check is made for any linked drafts. If a linked publication is still in draft status, publication is prevented
- Restrictions on linking to draft publications are removed from the API
- Restrictions on linking to draft publications the user is not an author on are present in the API 

---

### Checklist:

- [x] Local manual testing conducted
- [x] Automated tests added
- [ ] Documentation updated

---

### Tests:

API
Screenshot 2025-03-05 154017

E2E
Screenshot 2025-03-06 101805

### CODE REVIEW

File: api/src/components/coauthor/__tests__/updateAll.test.ts [hunk 1]
```diff
@@ -0,0 +1,337 @@
+import * as client from 'lib/client';
```
Comment (finlay-jisc) [inline api/src/components/coauthor/__tests__/updateAll.test.ts:1]: I had to add a test here but there were none so added tests for the endpoint generally.

File: api/src/components/link/controller.ts [hunk 2]
```diff
@@ -1,11 +1,200 @@
 import * as response from 'lib/response';
+import * as I from 'interface';
 import * as linkService from 'link/service';
+import * as publicationService from 'publication/service';
 
-import * as I from 'interface';
+export const canLinkBeCreatedBetweenPublicationTypes = (
```
Comment (finlay-jisc) [inline api/src/components/link/controller.ts:6]: Nothing new here, just moved this validation logic to the controller from the service because it seems a better home for it and helps to avoid circular imports.

File: api/src/components/publicationVersion/__tests__/updateStatus.test.ts [hunk 3]
```diff
@@ -378,4 +378,27 @@ describe('Update publication version status', () => {
 
         expect(updatePublicationVersionAttempt.status).toEqual(200);
     });
+
```
Comment (finlay-jisc) [inline api/src/components/publicationVersion/__tests__/updateStatus.test.ts:381]: The code as is satisfies these tests, just added them to ensure that the behaviour remains the same going forward as it's likely to be affected by following work on linking drafts.

File: e2e/tests/helpers/users.ts [hunk 4]
```diff
@@ -57,18 +57,19 @@ export const login = async (page: Page, browser: Browser, user: TestUser) => {
     await page.fill(PageModel.login.username, user.email);
     await page.fill(PageModel.login.password, user.password);
 
-    await Promise.all([
```
Comment (finlay-jisc) [inline e2e/tests/helpers/users.ts:60]: Unrelated to the work, but needed for the end to end tests to run.

**Review by karmseveeratjisc — APPROVED** [review]
phew!, looks good
